File: benchmarks/other_lang/bench_json.py
import sys
import time
import json

try:
    import discord  # noqa: F401
except Exception as exc:
    print("discord.py not available:", exc)
    sys.exit(1)


SAMPLE_JSON = '{"id":"123456789012345678","name":"test","value":42}'
SAMPLE_BYTES = SAMPLE_JSON.encode("utf-8")
MIN_WALL_SECONDS = 0.5


def _cpu_time():
    return time.process_time()


def bench(name, fn):
    iters = 1
    # warmup
    for _ in range(1000):
        fn()
    # calibrate iterations to reach MIN_WALL_SECONDS
    while True:
        start = time.perf_counter()
        for _ in range(iters):
            fn()
        wall = time.perf_counter() - start
        if wall >= MIN_WALL_SECONDS:
            break
        iters *= 2

    wall_start = time.perf_counter()
    cpu_start = _cpu_time()
    for _ in range(iters):
        fn()
    wall = time.perf_counter() - wall_start
    cpu = _cpu_time() - cpu_start
    ns_per_wall = (wall / iters) * 1e9
    ns_per_cpu = (cpu / iters) * 1e9
    print(f"{name:30s} {ns_per_wall:10.1f} ns {ns_per_cpu:10.1f} ns {iters:11d}")


def main():
    print("----------------------------------------------------------------")
    print("Benchmark                      Time             CPU   Iterations")
    print("----------------------------------------------------------------")
    bench("BM_JSON_Parse", lambda: json.loads(SAMPLE_JSON))
    bench("BM_JSON_Parse_Buffer", lambda: json.loads(SAMPLE_BYTES))

    obj = json.loads(SAMPLE_JSON)
    snowflake_str = obj["id"]

    def snowflake_op():
        int(snowflake_str)

    bench("BM_JSON_Get_Snowflake", snowflake_op)

    def serialize_op():
        payload = {"id": snowflake_str, "name": "test", "value": 42}
        json.dumps(payload, separators=(",", ":"))

    bench("BM_JSON_Mut_Serialize", serialize_op)


if __name__ == "__main__":
    main()

File: benchmarks/other_lang/bench_json.js
/* eslint-disable no-console */
const { performance } = require("perf_hooks");

let SnowflakeUtil = null;
try {
  ({ SnowflakeUtil } = require("discord.js"));
} catch (err) {
  console.error("discord.js not available. Install dependencies first:");
  console.error("  npm install");
  console.error("  # or");
  console.error("  bun install");
  process.exit(1);
}

const SAMPLE_JSON = '{"id":"123456789012345678","name":"test","value":42}';
const SAMPLE_BUF = Buffer.from(SAMPLE_JSON, "utf8");
const MIN_WALL_MS = 500;

function cpuUsageMs(start) {
  const usage = process.cpuUsage(start);
  return (usage.user + usage.system) / 1000;
}

function bench(name, fn) {
  let iters = 1;
  // warmup
  for (let i = 0; i < 1000; i += 1) fn();
  // calibrate
  while (true) {
    const start = performance.now();
    for (let i = 0; i < iters; i += 1) fn();
    const wall = performance.now() - start;
    if (wall >= MIN_WALL_MS) break;
    iters *= 2;
  }

  const wallStart = performance.now();
  const cpuStart = process.cpuUsage();
  for (let i = 0; i < iters; i += 1) fn();
  const wall = performance.now() - wallStart;
  const cpu = cpuUsageMs(cpuStart);
  const nsWall = (wall * 1e6) / iters;
  const nsCpu = (cpu * 1e6) / iters;
  console.log(
    `${name.padEnd(30)} ${nsWall.toFixed(1).padStart(10)} ns ` +
      `${nsCpu.toFixed(1).padStart(10)} ns ${String(iters).padStart(11)}`
  );
}

console.log("----------------------------------------------------------------");
console.log("Benchmark                      Time             CPU   Iterations");
console.log("----------------------------------------------------------------");
bench("BM_JSON_Parse", () => JSON.parse(SAMPLE_JSON));
bench("BM_JSON_Parse_Buffer", () => JSON.parse(SAMPLE_BUF.toString("utf8")));

const parsed = JSON.parse(SAMPLE_JSON);
const snowflake = parsed.id;

bench("BM_JSON_Get_Snowflake", () => {
  BigInt(snowflake);
});

bench("BM_JSON_Mut_Serialize", () => {
  const payload = { id: snowflake, name: "test", value: 42 };
  JSON.stringify(payload);
});
